src/dbtree.cpp

```diff
@@ -0,0 +1,343 @@
+    foreach (const QString &part, splitPath) {
+        currentSplitPath << part;
+        qDebug() << "getValue: currentSplitPath:" << currentSplitPath << "type:" << obj->type();
+
+        // make sure next level is an object
+        if (!obj->isObject()) {
+            qDebug() << "getValue() failed to traverse path:" << currentSplitPath;
+            return QMPointer<QMJsonValue>();
+        }
+
+        if (!obj->toObject()->contains(part)) {
+            qDebug() << "getValue() failed to traverse next path:" << currentSplitPath;
+            return QMPointer<QMJsonValue>();
+        }
+
+        obj = obj->toObject()->value(part);
+    }
+
+    qDebug() << "getValue: returning object";
+    return obj;
+}
+
+void DBTree::setValue(const QStringList &splitPath, QMPointer<QMJsonValue> value, bool skipFlush)
+{
+    QMPointer<QMJsonValue> obj = dbRoot;
+    auto db = lookupDb(splitPath);
+
+    // if it is top of tree, ignore
+    if (splitPath.length() == 0) {
+        qWarning("setValue: ignoring attempt to write to root");
+        return;
+    }
+
+    // split key from parent path values
+    QStringList parentList(splitPath);
+    QString key = parentList.takeLast();
+
+    // make tree as required
+    QStringList currentSplitPath;
+    auto currentDb = mainDb;
+    foreach (const QString &part, parentList) {
+        currentSplitPath << part;
+        qDebug() << "setValue: currentSplitPath:" << currentSplitPath << "type:" << obj->type();
+
+        if (!obj->toObject()->contains(part)) {
+            qDebug() << "setValue() inserting empty object";
+            currentDb->acquireWriteLock();
+            obj->toObject()->insert(part, QMPointer<QMJsonObject>(new QMJsonObject()));
+            currentDb->releaseWriteLock();
+        }
+
+        qDebug() << "setValue() getting value";
+        obj = obj->toObject()->value(part);
+
+        // make sure next level is an object
+        if (!obj->isObject()) {
+            qDebug() << "setValue() failed to traverse path:" << currentSplitPath << "type:" << obj->type();
+            return;
+        }
+
+        currentDb = lookupDb(currentSplitPath);
+    }
+
+    qDebug() << "setValue() inserting value:" << value;
+
+    currentDb->acquireWriteLock();
+    obj->toObject()->insert(key, value);
+    currentDb->releaseWriteLock();
+
+    // notify db to flush
+    if (!skipFlush) {
+        db->queueFlush();
+    }
+    return;
+}
+
+void DBTree::rmValue(const QStringList &splitPath)
+{
+    QMPointer<QMJsonValue> obj = dbRoot;
+    auto db = lookupDb(splitPath);
+
+    // if it is top of tree, ignore
+    if (splitPath.length() == 0) {
+        qWarning("rmValue: ignoring attempt to write to root");
+        return;
+    }
+
+    // split key from parent path values
+    QStringList parentList(splitPath);
+    QString key = parentList.takeLast();
+
+    // iterate through parent objects
+    QStringList currentSplitPath;
+    foreach (const QString &part, parentList) {
```

Comment:
same thing, I would do a grep for "foreach" and change to ranged for loops where you can